test/ProjectOrigin.Vault.Tests/WithdrawCertificatesTests.cs

```diff
@@ -0,0 +1,226 @@
+        var token = _jwtTokenIssuerFixture.GenerateToken(subject, name, scopes);
+        client.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", token);
+        return client;
+    }
+
+    public HttpClient CreateStampClient()
+    {
+        var client = new HttpClient();
+        client.BaseAddress = new Uri(_stampAndRegistryFixture.StampUrl);
+        return client;
+    }
+
+    [Fact]
+    public async Task WithdrawCertificate()
+    {
+        var registryName = _stampAndRegistryFixture.RegistryName;
+        var issuerArea = _stampAndRegistryFixture.IssuerArea;
+
+        var walletClient = CreateHttpClient(Guid.NewGuid().ToString(), Guid.NewGuid().ToString());
+        var wResponse = await walletClient.CreateWallet();
+        var weResponse = await walletClient.CreateWalletEndpoint(wResponse.WalletId);
+
+        var stampClient = CreateStampClient();
+        var rResponse = await stampClient.StampCreateRecipient(new CreateRecipientRequest
+        {
+            WalletEndpointReference = new StampWalletEndpointReferenceDto
+            {
+                Version = weResponse.WalletReference.Version,
+                Endpoint = weResponse.WalletReference.Endpoint,
+                PublicKey = weResponse.WalletReference.PublicKey.Export().ToArray()
+            }
+        });
+
+        var gsrn = Some.Gsrn();
+        var certificateId = Guid.NewGuid();
+        var icResponse = await stampClient.StampIssueCertificate(new CreateCertificateRequest
+        {
+            RecipientId = rResponse.Id,
+            RegistryName = registryName,
+            MeteringPointId = gsrn,
+            Certificate = new StampCertificateDto
+            {
+                Id = certificateId,
+                Start = DateTimeOffset.UtcNow.ToUnixTimeSeconds(),
+                End = DateTimeOffset.UtcNow.AddHours(1).ToUnixTimeSeconds(),
+                GridArea = issuerArea,
+                Quantity = 123,
+                Type = StampCertificateType.Production,
+                ClearTextAttributes = new Dictionary<string, string>
+                {
+                    { "fuelCode", Some.FuelCode },
+                    { "techCode", Some.TechCode }
+                },
+                HashedAttributes = new List<StampHashedAttribute>
+                {
+                    new () { Key = "assetId", Value = gsrn },
+                    new () { Key = "address", Value = "Some road 1234" }
+                }
+            }
+        });
+
+        await Task.Delay(TimeSpan.FromSeconds(15)); //wait for cert to be on registry and sent back to the wallet
+
+        var withdrawResponse = await stampClient.StampWithdrawCertificate(registryName, certificateId);
+
+        using (var connection = new NpgsqlConnection(_postgresFixture.GetConnectionString()))
+        {
+            var certificate = await connection.RepeatedlyQueryFirstOrDefaultUntil<Certificate>(
```

Comment:
What is this: RepeatedlyQueryFirstOrDefaultUntil ? :D

Reply:
We asynchronously query the database until the certificate has been withdrawn since it happens asynchronously :)